### ISSUE TITLE
Feature request: WebXRManager.getControllerGrip with filtered inputSource

### ISSUE THREAD

**Issue by Donitzo** [mrdoob/three.js#20348]
##### Description of the problem

This proposal is about how to select the correct controller with WebXR, which I have struggled with lately.

The current method of acquiring the controller grip and target-ray space objects are based on the controller index. When you call [xr.getController](https://threejs.org/docs/#api/en/renderers/webxr/WebXRManager.getController) or [xr.getControllerGrip](https://threejs.org/docs/#api/en/renderers/webxr/WebXRManager.getControllerGrip), a placeholder [WebXRController](https://github.com/mrdoob/three.js/blob/master/src/renderers/webxr/WebXRController.js) is created with the relevant spaces. When the _"inputsourceschange"_ event is fired the [XRSession.inputSources](https://developer.mozilla.org/en-US/docs/Web/API/XRSession/inputSources) are assigned to the controllers in the order they are declared in the inputSources array.

The problem is that at the time the WebXRController and placeholder objects are created, you do not have knowledge of to which hands these objects will be assigned. This means, as far as I can tell, that you can not fully setup the scene ahead of time with the correct hand models assigned to the correct placeholder objects, and that there is also no guarantee that the placeholders won't swap places in the future were the inputSources to change.

If available in your browser, [XRInputSource](https://developer.mozilla.org/en-US/docs/Web/API/XRInputSource) provides additional information about the current controller, such as targetRayMode and handedness. My proposed solution would be to allow the developer to filter which input sources are associated with different WebXRControllers based on the attributes of the XRInputSource. An additional index property could be provided to further select which of the matching input sources to use.

```
WebXRManager.prototype.getFilteredControllerGrip(filters) {...
```

Get grip associated with inputSource.handedness=="left", or the second input source,  or placeholder.
```
const handL = renderer.xr.getFilteredControllerGrip([{
  targetRayMode: "tracked-pointer",
  handedness: "left",
  matchIndex: 0,
}, {
  matchIndex: 1,
}]);
```
Get grip associated with inputSource.handedness=="right", or the first input source, or placeholder.
```
const handR = renderer.xr.getFilteredControllerGrip([{
  targetRayMode: "tracked-pointer",
  handedness: "right",
  matchIndex: 0,
}, {
  matchIndex: 0,
}]);
```

**Comment by mrdoob** [comment]
> ```
> renderer.xr.setDefaultHandedness("right");
> ```

Is this to support VR headsets that have only one controller?

**Comment by Donitzo** [comment]
> 
> 
> > ```
> > renderer.xr.setDefaultHandedness("right");
> > ```
> 
> Is this to support VR headsets that have only one controller?

Yes. It would be a method to select which hand the first input source should be assigned if handedness is missing entirely or has yet to be determined. 

Getting a hand by handedness makes sense when it exists, but there would have to be reasonable fallbacks too if it does not exist. I realize this is a rather complex issue, and it is important to not give the developer the impression that there exists more information than it does. In my opinion the best solution would be one in which the developer would both have to specify handedness or input index (ie, the old index) as a fallback.

**Comment by capnmidnight** [comment]
> I also do not know if handedness can change in an already established inputSource

I recall that some systems with ambidextrous controllers (primarily 3DOF standalones like Gear VR) had a "handedness preference" that could potentially be updated by the user mid-application usage.

In my own application, I have a bit of code somewhat like this:

````javascript
const motionControllers = new Map([
    ["left", new MotionControllerPointer("left")],
    ["right", new MotionControllerPointer("right")],
    ["none", new MotionControllerPointer("none")]
]);

for (let i = 0; i < 3; ++i) {
    setupController(i);
}

function setupController(ctrlIdx) {
    const ray = renderer.xr.getController(i);
    const grip = renderer.xr.getControllerGrip(i);
    const model = modelFactory.createControllerModel(ray);
    grip.add(model);

    let handedness = null;

    ray.addEventListener("connected", (evt) => {
        const session = renderer.xr.getSession();
        const device = session.inputSources[ctrlIdx];
        handedness = device.handedness;
        if (motionControllers.has(handedness)) {
            const mc = motionControllers.get(handedness);
            mc.inputSource = device;
            mc.controller = ray;
            mc.object.add(grip);
            mc.enabled = true;
        }
    });

    ray.addEventListener("disconnected", (evt) => {
        if (motionControllers.has(handedness)) {
            const mc = motionControllers.get(handedness);
            mc.enabled = false;
            mc.object.remove(grip);
            mc.controller = null;
            mc.inputSource = null;
        }
    });
};
````

In this way, I can pre-configure a controller and react to whichever hand the controller ends up being in.
